## Reduction operation: the sum of the numbers in the range [0, value)

In [7]:
import numpy as np
import sys

def reduc_operation(A):
    """Compute the sum of the elements of Array A in the range [0, value)."""
    s = 0
    for i in range(A.size):
        s += A[i]
    return s

# Secuencial

value = int(sys.argv[1])

X = np.random.rand(value)

# Para imprimir los pimeros valores del array

# print(X[0:12])

# Utilizando las operaciones mágicas de ipython

tiempo = %timeit -r 2 -o -q reduc_operation(X)

print("Time taken by reduction operation using a function:", tiempo)


print(f"And the result of the sum of numbers in the range [0, value) is: {reduc_operation(X)}\n")


# Utilizando numpy.sum()

tiempo = %timeit -r 2 -o -q np.sum(X)

print("Time taken by reduction operation using numpy.sum():", tiempo)

print("Now, the result using numpy.sum():", np.sum(X),"\n ")


# Utilizando numpy.ndarray.sum()

tiempo= %timeit -r 2 -o -q X.sum()

print("Time taken by reduction operation using numpy.ndarray.sum():", tiempo)

print("Now, the result using numpy.ndarray.sum():", X.sum())




Time taken by reduction operation using a function: 4.81 ms ± 26.2 µs per loop (mean ± std. dev. of 2 runs, 100 loops each)
And the result of the sum of numbers in the range [0, value) is: 25007.37808310279

Time taken by reduction operation using numpy.sum(): 13.9 µs ± 2.05 ns per loop (mean ± std. dev. of 2 runs, 100,000 loops each)
Now, the result using numpy.sum(): 25007.378083102678 
 
Time taken by reduction operation using numpy.ndarray.sum(): 11.8 µs ± 26.2 ns per loop (mean ± std. dev. of 2 runs, 100,000 loops each)
Now, the result using numpy.ndarray.sum(): 25007.378083102678


### EJERCICIO 3.3
#### APARTADO A

In [39]:
import numpy as np
import cupy as cp 
from cupyx.profiler import benchmark
import sys # Importamos los paquetes

value = int(sys.argv[1])
Y = cp.random.rand(value) # Creamos el array en la GPU

# Ahora medimos el tiempo de la suma con cp
tiempos_gpu = benchmark(cp.sum, (Y,), n_repeat=10, n_warmup=2)

# Calculamos la media de los tiempos y lo convertimos a ms
tiempo_medio = np.average(tiempos_gpu.gpu_times) * 1e3

print(f"El tiempo empleado en la suma con cupy es {tiempo_medio:.3f} milisegundos.")
print("El resultado de la suma es", cp.sum(Y))

El tiempo empleado en la suma con cupy es 0.045 milisegundos.
El resultado de la suma es 25046.111916890317


#### Resultados de la ejecución en bohr-gpu:

Ejecutando el script con el valor de 5000000 elementos

Time taken by reduction operation using a function: 536 ms ± 171 µs per loop (mean ± std. dev. of 2 runs, 1 loop each)
And the result of the sum of numbers in the range [0, value) is: 2499339.765069698

Time taken by reduction operation using numpy.sum(): 2.05 ms ± 636 ns per loop (mean ± std. dev. of 2 runs, 100 loops each)
Now, the result using numpy.sum(): 2499339.765069741

Time taken by reduction operation using numpy.ndarray.sum(): 2.04 ms ± 932 ns per loop (mean ± std. dev. of 2 runs, 100 loops each)
Now, the result using numpy.ndarray.sum(): 2499339.765069741

El tiempo empleado en la suma con cupy es 0.086 milisegundos.
El resultado de la suma es 2500216.390959031

Ejecutando el script con el valor de 50000000 elementos

Time taken by reduction operation using a function: 2.7 s ± 651 µs per loop (mean ± std. dev. of 2 runs, 1 loop each)
And the result of the sum of numbers in the range [0, value) is: 25001907.312928956

Time taken by reduction operation using numpy.sum(): 17.8 ms ± 921 ns per loop (mean ± std. dev. of 2 runs, 100 loops each)
Now, the result using numpy.sum(): 25001907.312925413

Time taken by reduction operation using numpy.ndarray.sum(): 17.8 ms ± 5.57 µs per loop (mean ± std. dev. of 2 runs, 100 loops each)
Now, the result using numpy.ndarray.sum(): 25001907.312925413

El tiempo empleado en la suma con cupy es 0.716 milisegundos.
El resultado de la suma es 24998738.856534127

Ejecutando el script con el valor de 500000000 elementos

Time taken by reduction operation using a function: 25.7 s ± 24.6 ms per loop (mean ± std. dev. of 2 runs, 1 loop each)
And the result of the sum of numbers in the range [0, value) is: 249991982.529075

Time taken by reduction operation using numpy.sum(): 179 ms ± 8.9 µs per loop (mean ± std. dev. of 2 runs, 10 loops each)
Now, the result using numpy.sum(): 249991982.52918264

Time taken by reduction operation using numpy.ndarray.sum(): 180 ms ± 50.6 µs per loop (mean ± std. dev. of 2 runs, 10 loops each)
Now, the result using numpy.ndarray.sum(): 249991982.52918264

El tiempo empleado en la suma con cupy es 6.839 milisegundos.
El resultado de la suma es 249992925.22264686

#### Conclusiones:
Podemos observar que el utilización del paquete CuPy y la ejecución en una cola con GPU disminuyen considerablemente el tiempo de la suma. La reducción de tiempo con respecto al resto de funciones aumenta conforme incrementamos el número de elementos a sumar.